consensus/core/src/commit_syncer.rs

```diff
@@ -0,0 +1,557 @@
+            match Self::fetch_once(inner.clone(), fetch_state.clone(), start, end).await {
+                Ok((commits, blocks)) => {
+                    info!(
+                        "Finished fetching commits from index {} (exclusive) to {} ...",
+                        start, end
+                    );
+                    return (end, commits, blocks);
+                }
+                Err(e) => {
+                    info!("Failed to fetch: {}", e);
+                }
+            }
+        }
+    }
+
+    async fn fetch_once(
+        inner: Arc<Inner<C>>,
+        fetch_state: Arc<Mutex<FetchState>>,
+        start: CommitIndex,
+        end: CommitIndex,
+    ) -> ConsensusResult<(Vec<TrustedCommit>, Vec<VerifiedBlock>)> {
+        const FETCH_COMMITS_TIMEOUT: Duration = Duration::from_secs(10);
+        const FETCH_BLOCKS_TIMEOUT: Duration = Duration::from_secs(60);
+        const FETCH_RETRY_BASE_INTERVAL: Duration = Duration::from_secs(1);
+        const FETCH_RETRY_INTERVAL_LIMIT: u32 = 30;
+        const MAX_RETRY_INTERVAL: Duration = Duration::from_secs(1);
+
+        let _timer = inner
+            .context
+            .metrics
+            .node_metrics
+            .commit_sync_fetch_once_latency
+            .start_timer();
+
+        // Find an available authority to fetch data.
+        let Some((available_time, target_authority, retries)) =
+            fetch_state.lock().available_authorities.pop_first()
+        else {
+            sleep(MAX_RETRY_INTERVAL).await;
+            return Err(ConsensusError::NoAvailableAuthorityToFetchCommits);
+        };
+        let now = Instant::now();
+        if now < available_time {
+            sleep(available_time - now).await;
+        }
+
+        // Fetch and verify commits.
+        let (serialized_commits, serialized_blocks) = match inner
+            .network_client
+            .fetch_commits(target_authority, start, end, FETCH_COMMITS_TIMEOUT)
+            .await
+        {
+            Ok(result) => {
+                let mut fetch_state = fetch_state.lock();
+                let now = Instant::now();
+                fetch_state
+                    .available_authorities
+                    .insert((now, target_authority, 0));
+                result
+            }
+            Err(e) => {
+                let mut fetch_state = fetch_state.lock();
+                let now = Instant::now();
+                fetch_state.available_authorities.insert((
+                    now + FETCH_RETRY_BASE_INTERVAL * retries.min(FETCH_RETRY_INTERVAL_LIMIT),
+                    target_authority,
+                    retries.saturating_add(1),
+                ));
+                return Err(e);
+            }
+        };
+
+        let commits = inner.verify_commits(
+            target_authority,
+            start,
+            end,
+            serialized_commits,
+            serialized_blocks,
+        )?;
+
+        // Fetch blocks and verify with digests.
```

Comment:
I think I am misunderstanding or missing something. It seems like we are getting the blocks & commits in the `fetch_commits` call and then verifying the blocks in `verify_commits`. But then we are fetching the same blocks again and doing some light digest verification? If that is true, can you explain why we need to do it twice?

Reply:
Updated the comments in a few places and let me know where else they can be improved. The returned blocks from `fetch_commit()` RPC only contains blocks certifying the last returned commit. They are not the blocks referenced by the commit. A separate step is needed to fetch the blocks referenced by the commits. Since the commits are certified, the referenced blocks can be considered verified as well.

The reasons for the separate block fetching step are:
1. Blocks can be large, so it seems better to reuse the chunking & size limiting logic in `fetch_blocks()`.
2. In future, we can implement logic where after commits are successfully fetched and certified, only block fetching are retried if failed.